modules/core/src/layers/xviz-layer.js

```diff
@@ -289,120 +406,69 @@ export default class XVIZLayer extends CompositeLayer {
-            getPosition: f => f.vertices || f.center,
-            updateTriggers: deepExtend(updateTriggers, {
-              getFillColor: {useSemanticColor: this.props.useSemanticColor}
-            })
-          })
-        );
-
-      case 'pointcloud':
-        return new PointCloudLayer(
-          forwardProps,
-          layerProps,
-          this.getSubLayerProps({
-            id: 'pointcloud',
-            data: {
-              length: data[0].points.length / 3,
-              attributes: {
-                getPosition: data[0].points,
-                getColor: data[0].colors
-              }
-            },
-            vehicleRelativeTransform: this.props.vehicleRelativeTransform,
-            getPosition: p => p
-          })
-        );
-
-      case 'path':
-        return new PathLayer(
-          forwardProps,
+    // multiple layers?
+    const customXvizLayerMatch = this.props.customXVIZLayers.find(({streamMatch}) =>
+      streamMatch(streamName, streamMetadata)
+    );
+
+    if (customXvizLayerMatch) {
+      let primitiveLayerProps = {};
+      const parentLayerHandler = LAYER_HANDLERS[customXvizLayerMatch.extendPrimitive];
+      if (parentLayerHandler) {
+        primitiveLayerProps = parentLayerHandler.getLayerTypeProps({
+          xvizLayerProps: this.props,
           layerProps,
-          this.getSubLayerProps({
-            id: 'path',
-            data,
-            getPath: f => f.vertices,
-            updateTriggers: deepExtend(updateTriggers, {
-              getColor: {useSemanticColor: this.props.useSemanticColor}
-            })
-          })
-        );
-
-      case 'stadium':
-        return new PathLayer(
-          forwardProps,
-          layerProps,
-          this.getSubLayerProps({
-            id: 'stadium',
-            data,
-            getPath: f => [f.start, f.end],
-            rounded: true,
-            updateTriggers: deepExtend(updateTriggers, {
-              getColor: {useSemanticColor: this.props.useSemanticColor}
-            })
-          })
-        );
+          state: this.state
+        });
+      }
 
-      case 'polygon':
-        return new PolygonLayer(
-          forwardProps,
-          layerProps,
-          this.getSubLayerProps({
-            id: 'polygon',
-            opacity: this.props.opacity || 1,
-            data,
-            lightSettings,
-            wireframe: layerProps.stroked,
-            getPolygon: f => f.vertices,
-            updateTriggers: deepExtend(updateTriggers, {
-              getLineColor: {useSemanticColor: this.props.useSemanticColor},
-              getFillColor: {useSemanticColor: this.props.useSemanticColor}
-            })
+      const LayerClass = customXvizLayerMatch.layerClass || parentLayerHandler?.layerClass;
+      return new LayerClass(
+        forwardProps,
+        layerProps,
+        this.getSubLayerProps({
+          id: customXvizLayerMatch.id,
```

Comment:
what about storing the id in the layerHandler, and simply appending the customLayer.id?  Right now we have an `id` in the custom layer, but the getSublayerProps also sets the `id` used here, which is precisely the concatenated original + customLayer.id.